dev/TabView/TabView.cpp

```diff
@@ -0,0 +1,260 @@
+#include "ResourceAccessor.h"
+#include "SharedHelpers.h"
+
+static constexpr double c_tabMinimumWidth = 48.0;
+static constexpr double c_tabMaximumWidth = 200.0;
+
+static constexpr wstring_view c_tabViewItemMinWidthName{ L"TabViewItemMinWidth"sv };
+static constexpr wstring_view c_tabViewItemMaxWidthName{ L"TabViewItemMaxWidth"sv };
+
+// TODO: what is the right number and should this be customizable?
+static constexpr double c_scrollAmount = 50.0;
+
+TabView::TabView()
+{
+    __RP_Marker_ClassById(RuntimeProfiler::ProfId_TabView);
+
+    SetDefaultStyleKey(this);
+
+    Loaded({ this, &TabView::OnLoaded });
+    SelectionChanged({ this, &TabView::OnSelectionChanged });
+    SizeChanged({ this, &TabView::OnSizeChanged });
+}
+
+void TabView::OnApplyTemplate()
+{
+    winrt::IControlProtected controlProtected{ *this };
+
+    m_tabContentPresenter.set(GetTemplateChildT<winrt::ContentPresenter>(L"TabContentPresenter", controlProtected));
+    m_scrollViewer.set(GetTemplateChildT<winrt::FxScrollViewer>(L"ScrollViewer", controlProtected));
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        m_scrollViewerLoadedRevoker = scrollViewer.Loaded(winrt::auto_revoke, { this, &TabView::OnScrollViewerLoaded });
+    }
+}
+
+void TabView::OnPropertyChanged(const winrt::DependencyPropertyChangedEventArgs& args)
+{
+    winrt::IDependencyProperty property = args.Property();
+
+    if (property == s_TabWidthModeProperty)
+    {
+        UpdateTabWidths();
+    }
+}
+
+winrt::AutomationPeer TabView::OnCreateAutomationPeer()
+{
+    return winrt::make<TabViewAutomationPeer>(*this);
+}
+
+void TabView::OnLoaded(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    UpdateTabContent();
+}
+
+void TabView::OnScrollViewerLoaded(const winrt::IInspectable& sender, const winrt::RoutedEventArgs& args)
+{
+    if (auto scrollViewer = m_scrollViewer.get())
+    {
+        m_scrollDecreaseButton.set(SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollDecreaseButton").as<winrt::RepeatButton>());
+        if (auto decreaseButton = m_scrollDecreaseButton.get())
+        {
+            m_scrollDecreaseClickRevoker = decreaseButton.Click(winrt::auto_revoke, { this, &TabView::OnScrollDecreaseClick });
+        }
+
+        m_scrollIncreaseButton.set(SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollIncreaseButton").as<winrt::RepeatButton>());
+        if (auto increaseButton = m_scrollIncreaseButton.get())
+        {
+            m_scrollIncreaseClickRevoker = increaseButton.Click(winrt::auto_revoke, { this, &TabView::OnScrollIncreaseClick });
+        }
+
+        if (SharedHelpers::IsRS2OrHigher())
+        {
+            if (auto scrollContentPresenter = SharedHelpers::FindInVisualTreeByName(scrollViewer, L"ScrollContentPresenter").as<winrt::ScrollContentPresenter>())
+            {
+                scrollContentPresenter.TabFocusNavigation(winrt::KeyboardNavigationMode::Once);
+            }
```

Comment:
We should be able to use conditional markup in the XAML file to set this instead.

Reply:
Not without duplicating my xaml file into rs1 and rs2+ files. I would rather have one line of code than have to care for two files that are identical except for one line.

Reply:
CanHua made this possible now -- there's a script that processes the conditionals now. You might need to set a thing to get it to happen, but the conditionals can work down to RS1.

Reply:
Is there an example I can follow?

Reply:
Search for UseNonstandardConditionalXAML in vcxitems:

```xml
    <Page Include="$(MSBuildThisFileDirectory)AppBarButton_themeresources.xaml">
      <Version>RS1</Version>
      <Type>ThemeResources</Type>
      <UseNonstandardCondtionalXAML>true</UseNonstandardCondtionalXAML>
    </Page>
```

---
In reply to: [280932133](https://github.com/Microsoft/microsoft-ui-xaml/pull/629#discussion_r280932133) [](ancestors = 280932133)

Reply:
I see that, but setting it alone doesn't seem to work. Is that only for ThemeResources?

Reply:
I don't think so, but @licanhua should be able to advise.

Reply:
UseNonstandardCondtionalXAML flag only impact tests right now. It just tell test pages that I'm not using offical conditional xaml, and please exclude me from the tests.
There are bunch of examples in dev/CommonStyles.

You may need to add xmlns:contract5Present,
then set it like this:
  <contract5Present:Setter Property="UseSystemFocusVisuals" Value="{ThemeResource IsApplicationFocusVisualKindReveal}" />



---
In reply to: [281264015](https://github.com/Microsoft/microsoft-ui-xaml/pull/629#discussion_r281264015) [](ancestors = 281264015)

Reply:
I had other build issues, apparently. Syncing to master once more has got it working!